### PR TITLE
#164069183 create articles

### PR DESCRIPTION
**Pull request by wasibani-roy**
**What does this PR do?**
This PR enables the user of our system to create articles on our system

**Description of Task to be completed?**

- user should be able to create articles and view them
- user should be able to edit an article they have created
- User should be able to delete an article
- The navigation bar should change when someone is not logged in
- The user should be able to log out of our application

**How should this be manually tested?**

- Clone the repo and cd into `ah-aquaman-frontend`
- checkout the branch `ft-create-article-164069183`
- Run the commands below to install packages and run the server
         `npm install`
        ` npm start`
- Navigate to the following routes
      - `/create_article` to create an article
      - `/article/<articleSlug>` to view an article
      - `/editor/<articleSlug>` to edit an article

**What are the relevant pivotal tracker stories?**

- [#164069183](https://www.pivotaltracker.com/story/show/164069183)

**Screenshot**

**Create an article**
Screenshot 2019-04-17 at 19 18 23

**View an article**
Screenshot 2019-04-17 at 19 04 16

**View all articles**
Screenshot 2019-04-17 at 19 04 40

### CODE REVIEW

File: src/__tests__/actions/createArticle/__test__/createArticle.test.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
+import thunk from "redux-thunk";
+import configureStore from "redux-mock-store";
+import fetchMock from "fetch-mock";
+import { CreateArticle } from "../../../../actions/articleActions/createArticle";
+import { ArticleActionTypes } from "../../../../actions/types";
+
+
```
Comment (joelethan) [inline src/__tests__/actions/createArticle/__test__/createArticle.test.js]: @wasibani-roy This redundant line should be removed

File: src/__tests__/actions/createArticle/__test__/createArticle.test.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        type: ArticleActionTypes.CREATE_ARTICLE_SUCCESS,
+        payload: {
+          title: "test article",
+          description: "this is a test article",
+          body: "Test article by a testuser",
+        },
+      },
+    ];
+
+    const store = mockStore();
```
Comment (joelethan) [inline src/__tests__/actions/createArticle/__test__/createArticle.test.js]: Could you have a single store in describe to remove code duplications?

File: src/__tests__/actions/deleteArticle/__test__/deleteArticle.test.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
+import thunk from "redux-thunk";
+import configureStore from "redux-mock-store";
+import fetchMock from "fetch-mock";
+import { DeleteArticle } from "../../../../actions/articleActions/createArticle";
+import { ArticleActionTypes } from "../../../../actions/types";
+
+
```
Comment (joelethan) [inline src/__tests__/actions/deleteArticle/__test__/deleteArticle.test.js]: Remove this redundant line.

File: src/__tests__/actions/editArticle/__test__/editArticle.test.js [hunk 4]
```diff
@@ -0,0 +1,93 @@
+import thunk from "redux-thunk";
+import configureStore from "redux-mock-store";
+import fetchMock from "fetch-mock";
+import { EditArticle } from "../../../../actions/articleActions/createArticle";
+import { ArticleActionTypes } from "../../../../actions/types";
+
+
```
Comment (joelethan) [inline src/__tests__/actions/editArticle/__test__/editArticle.test.js]: Remove this redundant line.

File: src/__tests__/reducers/articleReducer/__test__/articleReducer.test.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+      loading: false,
+    };
+    const action = {
+      type: ArticleActionTypes.ARTICLE_FETCH_SUCCESSFUL,
+      payload: {},
+    };
+    expect(ArticleReducer({}, action)).toEqual(expectedState);
+  });
+
+
```
Comment (joelethan) [inline src/__tests__/reducers/articleReducer/__test__/articleReducer.test.js]: Remove this redundant line.

File: src/__tests__/views/createArticleView/__test__/createArticleView.test.js [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+  isSuccessful: false,
+  errors: {
+    title: "incorrect title",
+    description: "incorrect description",
+    body: "field can not be blank",
+  },
+  CreateArticle: jest.fn(),
+};
+
+
```
Comment (joelethan) [inline src/__tests__/views/createArticleView/__test__/createArticleView.test.js]: Remove this redundant line.

File: src/__tests__/views/editArticleView/__test__/editArticleView.test.js [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+  errors: {
+    title: "error in title",
+    description: "error in description",
+    body: "error in body",
+  },
+  EditArticle: jest.fn(),
+  getSingleArticle: jest.fn(),
+};
+
+
```
Comment (joelethan) [inline src/__tests__/views/editArticleView/__test__/editArticleView.test.js:34]: Remove this redundant line.

File: src/__tests__/views/loginView/__test__/loginView.test.js [hunk 8]
```diff
@@ -90,4 +91,19 @@ describe("LoginView", () => {
...
+        isSuccessful: true,
+      },
+    };
+    expect(mapStateToProps(state)).toEqual({
+      token: "",
+      errors: {},
+      isSuccessful: true,
+    });
+  });
+});
```
Comment (joelethan) [inline src/__tests__/views/loginView/__test__/loginView.test.js]: Please add a new line at the end of this file.

File: src/__tests__/views/logoutView/__test__/logoutView.test.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+    push: jest.fn(),
+  },
+};
+
+describe("Logout view test", () => {
+  it("should render without crashing", () => {
+    const wrapper = shallow(<LogoutPage {...props} />);
+    expect(wrapper).toMatchSnapshot();
+  });
+});
```
Comment (joelethan) [inline src/__tests__/views/logoutView/__test__/logoutView.test.js]: Please add a new line at the end of this file.

File: src/actions/articleActions/createArticle/index.js [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+      } else {
+        dispatch({
+          type: ArticleActionTypes.EDIT_ARTICLE_SUCCESS,
+          payload: data,
+        });
+      }
+    },
+  );
+
+
```
Comment (joelethan) [inline src/actions/articleActions/createArticle/index.js]: Remove this redundant line.

File: src/components/article/ImageBanner/index.js [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+  render() {
+    return (
+      <React.Fragment>
+        <section className="image-banner">
+          <img src={Banner} alt="" />
+          <div className="container">
+            <div className="banner-content text-center">
+              <h2>{this.props.info}</h2>
+              <p>
+Home
```
Comment (joelethan) [inline src/components/article/ImageBanner/index.js]: Please fix the tabs or number of spaces in this line.

File: src/components/article/articleComponent/index.js [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+          </div>
+          <div className="col-md-6">
+            <DeleteButton
+              onClick={props.onClick}
+              slug={props.slug}
+              username={props.article.author.username}
+            />
+          </div>
+        </div>
+
```
Comment (joelethan) [inline src/components/article/articleComponent/index.js]: Please fix these redundant lines

File: src/components/article/editArticleButton/index.js [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+            </span>
+          </Link>
+        </div>
+      );
+    }
+    return null;
+  }
+}
+
+
```
Comment (joelethan) [inline src/components/article/editArticleButton/index.js]: Remove this redundant line.

File: src/views/editArticleView/index.js [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+          </div>
+          <ToastContainer />
+        </div>
+      </React.Fragment>
+      
+    );
+  }
+}
+
+
```
Comment (joelethan) [inline src/views/editArticleView/index.js]: Remove this redundant line.

File: .codeclimate.yml [hunk 15]
```diff
@@ -5,3 +5,7 @@ checks:
   method-lines:
     config:
       threshold: 80
+  similar-code:
```
Comment (CryceTruly) [inline .codeclimate.yml:8]: Great work @wasibani-roy, What are these lines meant to do?

File: public/index.html [hunk 16]
```diff
@@ -16,6 +19,7 @@
     <div id="root">
 
     </div>
+    <script src="https://cdnjs.cloudflare.com/ajax/libs/toastr.js/latest/toastr.min.js"></script>
```
Comment (CryceTruly) [inline public/index.html:22]: Great work on using the CDN link as it loads the resources faster

**Review by joelethan — CHANGES_REQUESTED** [review]

File: src/__tests__/actions/createArticle/__test__/createArticle.test.js [hunk 17]
```diff
@@ -0,0 +1,90 @@
...
+
+    return store.dispatch(CreateArticle(data)).then(() => {
+      expect(store.getActions()).toEqual(expectedAction);
+    });
+  });
+
+  it("tests faliure to create an article", () => {
```
Comment (CryceTruly) [inline src/__tests__/actions/createArticle/__test__/createArticle.test.js]: Hey @wasibani-roy ,Check grammar here

File: src/__tests__/actions/editArticle/__test__/editArticle.test.js [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+const slug = "article-slug";
+
+describe("Edit an article test", () => {
+  beforeEach(() => {
+    fetchMock.restore();
+    localStorage.clear();
+  });
+
+  it("tests update an article", () => {
```
Comment (CryceTruly) [inline src/__tests__/actions/editArticle/__test__/editArticle.test.js]: Hey @wasibani-roy, Please grammar here, let the tests and test suite descriptions be as descriptive as possible.
Examples of this would be like `it('should update an article')`

File: src/__tests__/actions/editArticle/__test__/editArticle.test.js [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+    ];
+
+    const store = mockStore();
+
+    return store.dispatch(EditArticle(slug, data)).then(() => {
+      expect(store.getActions()).toEqual(expectedAction);
+    });
+  });
+
+  it("tests faliure to update an article", () => {
```
Comment (CryceTruly) [inline src/__tests__/actions/editArticle/__test__/editArticle.test.js]: You can make it more descriptive.

File: src/__tests__/actions/getSingleArticle/__test__/getSingleArticle.test.js [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+
+    const store = mockStore();
+
+    return store.dispatch(getSingleArticle(slug)).then(() => {
+      expect(store.getActions()).toEqual(expectedActions);
+    });
+  });
+
+
+  it("should dispatch ARTICLE_FETCH_NOT_FOUND action", () => {
```
Comment (CryceTruly) [inline src/__tests__/actions/getSingleArticle/__test__/getSingleArticle.test.js:43]: Good test description 👍

File: src/__tests__/components/navbarComponent/__test__/navbarComponent.test.js [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+      hidden: false,
+    },
+    history: {
+      push: jest.fn(),
+    },
+  };
+
+  it("component should mount", () => {
+    const wrapper = mount(<Router><Navbar {...props} /></Router>);
+    console.log(wrapper.instance().props.children.props);
```
Comment (CryceTruly) [inline src/__tests__/components/navbarComponent/__test__/navbarComponent.test.js]: Hey @wasibani-roy , You could eliminate this console log, when you get done testing on your local environment.

File: src/__tests__/reducers/articleReducer/__test__/articleReducer.test.js [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+      article: {},
+      loading: false,
+    };
+    const action = {
+      type: ArticleActionTypes.ARTICLE_FETCH_SUCCESSFUL,
+      payload: {},
+    };
+    expect(ArticleReducer({}, action)).toEqual(expectedState);
+  });
+
```
Comment (CryceTruly) [inline src/__tests__/reducers/articleReducer/__test__/articleReducer.test.js:18]: Leave only one line here,for consistency

File: src/__tests__/views/articleDetailView/__test__/articleDetailView.test.js [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+          slug: "article-slug",
+        },
+      },
+      getSingleArticle: jest.fn(),
+    };
+    const wrapper = new ArticleDetailView(props);
+    wrapper.state = { notFound: true };
+    expect(wrapper.render()).toMatchSnapshot();
+  });
+
```
Comment (CryceTruly) [inline src/__tests__/views/articleDetailView/__test__/articleDetailView.test.js:47]: Leave a single line here.

File: src/__tests__/views/loginView/__test__/loginView.test.js (hunk 8 above)
Comment (CryceTruly) [inline src/__tests__/views/loginView/__test__/loginView.test.js]: Hey @wasibani-roy ,Please add a line here

File: src/__tests__/views/logoutView/__test__/logoutView.test.js (hunk 9 above)
Comment (CryceTruly) [inline src/__tests__/views/logoutView/__test__/logoutView.test.js]: Hey @wasibani-roy ,You should add a new line here

File: src/components/article/ImageBanner/index.js [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+                <i className="fa fa-angle-right" /> 
+                {" "}
+                {this.props.page}
```
Comment (CryceTruly) [inline src/components/article/ImageBanner/index.js:17]: Hey @wasibani-roy please help  clarify on importance of this expression

**Review by CryceTruly — CHANGES_REQUESTED** [review]

**Review by CryceTruly — COMMENTED** [review]
Great Progress and Nice implementation @wasibani-roy,
We can improve by adding a custom alert model to show alert options.

File: src/views/editArticleView/index.js (hunk 14 above)
Comment (wasibani-roy) [reply src/views/editArticleView/index.js]: Corrected

File: src/components/article/editArticleButton/index.js (hunk 13 above)
Comment (wasibani-roy) [reply src/components/article/editArticleButton/index.js]: Resolved

File: src/components/article/articleComponent/index.js (hunk 12 above)
Comment (wasibani-roy) [reply src/components/article/articleComponent/index.js]: Changes made

File: src/components/article/ImageBanner/index.js (hunk 11 above)
Comment (wasibani-roy) [reply src/components/article/ImageBanner/index.js]: corrected

File: src/actions/articleActions/createArticle/index.js (hunk 10 above)
Comment (wasibani-roy) [reply src/actions/articleActions/createArticle/index.js]: Thank you i have made changes

File: src/__tests__/reducers/articleReducer/__test__/articleReducer.test.js (hunk 5 above)
Comment (wasibani-roy) [reply src/__tests__/reducers/articleReducer/__test__/articleReducer.test.js]: Done

File: src/__tests__/actions/editArticle/__test__/editArticle.test.js (hunk 4 above)
Comment (wasibani-roy) [reply src/__tests__/actions/editArticle/__test__/editArticle.test.js]: corrected

File: src/__tests__/actions/deleteArticle/__test__/deleteArticle.test.js (hunk 3 above)
Comment (wasibani-roy) [reply src/__tests__/actions/deleteArticle/__test__/deleteArticle.test.js]: Thank you i have made corrections

File: src/__tests__/actions/createArticle/__test__/createArticle.test.js (hunk 2 above)
Comment (wasibani-roy) [reply src/__tests__/actions/createArticle/__test__/createArticle.test.js]: I have moved the store out

File: src/__tests__/actions/createArticle/__test__/createArticle.test.js (hunk 1 above)
Comment (wasibani-roy) [reply src/__tests__/actions/createArticle/__test__/createArticle.test.js]: Okay made changes

File: .codeclimate.yml (hunk 15 above)
Comment (wasibani-roy) [reply .codeclimate.yml:8]: This line handles sections that are similar in look but have different implementations. So code climate will not dwell to much on them.

File: public/index.html (hunk 16 above)
Comment (wasibani-roy) [reply public/index.html:22]: Thank you

File: src/components/article/ImageBanner/index.js (hunk 24 above)
Comment (wasibani-roy) [reply src/components/article/ImageBanner/index.js:17]: This just helps me put an empty space before i display the page

**Review by joelethan — APPROVED** [review]

**Comment by Peace-Apple** [comment]
Great work @wasibani-roy, I have checked out the functionality and it works fine, I can create, view edit and delete, and good work on adding logout meanwhile rebase to resolve the conflicts your branch is having.

File: .codeclimate.yml (hunk 15 above)
Comment (CryceTruly) [reply .codeclimate.yml:8]: okay Roy

**Review by Peace-Apple — APPROVED** [review]

**Review by CryceTruly — APPROVED** [review]
Great work @wasibani-roy

File: src/__tests__/components/deleteArticleButton/__test__/deleteArticleButton.test.js [hunk 25]
```diff
@@ -0,0 +1,68 @@
...
+const props = {
+  isArticleDeleted: false,
+  slug: "article-slug",
+  history: {
+    push: jest.fn(),
+  },
+  DeleteArticle: jest.fn(),
+};
+
+global.confirm = () => true;
```
Comment (AmosWels) [inline src/__tests__/components/deleteArticleButton/__test__/deleteArticleButton.test.js:14]: does this use the default `js` confirm?

File: src/__tests__/components/navbarComponent/__test__/navbarComponent.test.js [hunk 26]
```diff
@@ -0,0 +1,23 @@
...
+import { shallow, mount } from "enzyme";
+import { Provider } from "react-redux";
+import configureStore from "redux-mock-store";
+import Navbar from "../../../../components/article/navbarComponent/Navbar";
+
+const initialState = {};
+const mockStore = configureStore();
+const store = mockStore(initialState);
+
+describe("Navbar not logged in test", () => {
```
Comment (AmosWels) [inline src/__tests__/components/navbarComponent/__test__/navbarComponent.test.js]: add a hyphen to improve the grammar in this title

File: src/__tests__/components/profile/Profile.test.js [hunk 27]
```diff
@@ -21,7 +22,7 @@ describe("profile component tests", () => {
   });
 
   it("should render when mounted", () => {
-    const wrapper = mount(<Provider store={store}><Profile /></Provider>);
+    const wrapper = mount(<Provider store={store}><Router><Profile /></Router></Provider>);
```
Comment (AmosWels) [inline src/__tests__/components/profile/Profile.test.js:25]: why did you add `Router` on the profile component

File: src/__tests__/routes/__test__/protectedRoute.test.js [hunk 28]
```diff
@@ -20,4 +22,23 @@ describe("protected routes", () => {
...
   });
+
+  it("should protectedRoute component should render without crashing", () => {
+    const wrapper = shallow(<ProtectedRoute />);
+    expect(wrapper).toMatchSnapshot();
+  });
+
+  it("test valid token at login but expired", () => {
+    const falseToken = Authenticate(
+      "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJlbWFpbCI6ImtpbWJzaW1vbjJAZ21haWwuY29tIiwiZXhwIjoxNTQyMDYxMjQ3LCJ1c2VybmFtZSI6ImtpbWJ1Z3AifQ.yz9jMoPBnKy9Tko_hhCwakNwZkeRLD4nc13A3LzyJbw",
```
Comment (AmosWels) [inline src/__tests__/routes/__test__/protectedRoute.test.js:33]: do you need this line here

**Review by AmosWels — CHANGES_REQUESTED** [review]
Goodwork @wasibani-roy. Please format your PR description to indicate noteworthy areas like the branch name, the specifc commands to run etc. 

Also make sure your intended formats are working.

Please add screen shots to your PR description.

Also check that all the unnecessary unused files are removed except if you are using them such that your `eslint` issues are resolved

**Review by joelethan — CHANGES_REQUESTED** [review]
@wasibani-roy In the commit message, add a new line to separate the body and the footer.

File: src/__tests__/components/deleteArticleButton/__test__/deleteArticleButton.test.js (hunk 25 above)
Comment (wasibani-roy) [reply src/__tests__/components/deleteArticleButton/__test__/deleteArticleButton.test.js:14]: Yes, it uses the js alert box. Here I set it to true so meaning that in this case I basically click the ok button in the alert box. so it simulates that action for me in the test

File: src/__tests__/components/profile/Profile.test.js (hunk 27 above)
Comment (wasibani-roy) [reply src/__tests__/components/profile/Profile.test.js:25]: The Router is added because in our Profile component we are using a Link prop and this prop must be wrapped in a Router,  I do that in the tests and that way the Component is able to mount

File: src/__tests__/routes/__test__/protectedRoute.test.js (hunk 28 above)
Comment (wasibani-roy) [reply src/__tests__/routes/__test__/protectedRoute.test.js:33]: Yes i need to pass in a token that has expired so that it simulates how the component would react when a users token is no longer valid

**Review by joelethan — APPROVED** [review]

**Review by AmosWels — APPROVED** [review]
